yt_dlp/downloader/youtube_dl_from_start.py

```diff
@@ -0,0 +1,89 @@
+from __future__ import unicode_literals
+
+import re
+from time import sleep
+
+from .dash import DashSegmentsFD
+from ..extractor.youtube import YoutubeIE
+
+from ..utils import (
+    traverse_obj,
+    urljoin,
+    time_millis,
+)
+
+
+class YoutubeDlFromStartDashFD(DashSegmentsFD):
+    """
+    Download YouTube live from the start, to the end. For DASH formats.
+    """
+
+    FD_NAME = 'ytlivestartdash'
+
+    @staticmethod
+    def _manifest_fragments(ie: YoutubeIE, mpd_url, stream_number, begin_index=0, fetch_span=5000):
+        known_idx = begin_index + 1
+        no_fragment_count = 0
+        prev_dl = time_millis()
+        while True:
+            if no_fragment_count > 5:
+                return
+            fmts, _ = ie._extract_mpd_formats_and_subtitles(
+                mpd_url, None, note=False, errnote=False, fatal=False)
+            if not fmts:
+                no_fragment_count += 1
+                continue
+            fmt_info = next(x for x in fmts if x['manifest_stream_number'] == stream_number)
+            fragments = fmt_info['fragments']
+            fragment_base_url = fmt_info['fragment_base_url']
+            assert fragment_base_url
+
+            last_seq = int(re.search(r'(?:/|^)sq/(\d+)', fragments[-1]['path']).group(1)) + 1
+            if begin_index < 0 and known_idx <= 0:
+                # skip from the start when it's negative value
+                known_idx = last_seq + begin_index
+            for idx in range(known_idx, last_seq):
+                yield {
+                    'frag_index': idx - 1,
+                    'index': idx - 1,
+                    'url': urljoin(fragment_base_url, 'sq/%d' % idx),
+                }
```

Comment:
As I understand, this URL fetching and the below max duration check are the only YouTube specific code. An I correct?

Reply:
yes. YouTube-specific code.